reactos/drivers/usb/usbhub_new/usbhub.c

```diff
@@ -0,0 +1,5074 @@
+    {
+        KeSetEvent(&HubExtension->ResetEvent,
+                   EVENT_INCREMENT,
+                   FALSE);
+
+        if (HubExtension->HubFlags & USBHUB_FDO_FLAG_DEFER_CHECK_IDLE)
+        {
+            USBH_CheckHubIdle(HubExtension);
+        }
+    }
+}
+
+NTSTATUS
+NTAPI
+USBH_ChangeIndication(IN PDEVICE_OBJECT DeviceObject,
+                      IN PIRP Irp,
+                      IN PVOID Context)
+{
+    PUSBHUB_FDO_EXTENSION HubExtension;
+    USBD_STATUS UrbStatus;
+    BOOLEAN IsErrors = FALSE;
+    PUSBHUB_IO_WORK_ITEM HubWorkItem;
+    PUSBHUB_STATUS_CHANGE_CONTEXT HubWorkItemBuffer;
+    USHORT NumPorts;
+    USHORT Port;
+    NTSTATUS Status;
+    PVOID Bitmap;
+    ULONG BufferLength;
+
+    HubExtension = Context;
+    UrbStatus = HubExtension->SCEWorkerUrb.Hdr.Status;
+
+    DPRINT("USBH_ChangeIndication: IrpStatus - %x, UrbStatus - %x, HubFlags - %lX\n",
+           Irp->IoStatus.Status,
+           UrbStatus,
+           HubExtension->HubFlags);
+
+    if (NT_ERROR(Irp->IoStatus.Status) || USBD_ERROR(UrbStatus) ||
+       (HubExtension->HubFlags & USBHUB_FDO_FLAG_DEVICE_FAILED) ||
+       (HubExtension->HubFlags & USBHUB_FDO_FLAG_DEVICE_STOPPING))
+    {
+        HubExtension->RequestErrors++;
+
+        IsErrors = TRUE;
+
+        KeSetEvent(&HubExtension->StatusChangeEvent,
+                   EVENT_INCREMENT,
+                   FALSE);
+
+        if (HubExtension->HubFlags & USBHUB_FDO_FLAG_DEVICE_STOPPING ||
+            HubExtension->HubFlags & USBHUB_FDO_FLAG_DEVICE_FAILED ||
+            HubExtension->RequestErrors > USBHUB_MAX_REQUEST_ERRORS ||
+            Irp->IoStatus.Status == STATUS_DELETE_PENDING)
+        {
+            DPRINT("USBH_ChangeIndication: HubExtension->RequestErrors - %x\n",
+                   HubExtension->RequestErrors);
+
+            return STATUS_MORE_PROCESSING_REQUIRED;
+        }
+
+        DPRINT("USBH_ChangeIndication: HubExtension->RequestErrors - %x\n",
+               HubExtension->RequestErrors);
+    }
+    else
+    {
+        HubExtension->RequestErrors = 0;
+    }
+
+    BufferLength = sizeof(USBHUB_STATUS_CHANGE_CONTEXT) +
+                   HubExtension->SCEBitmapLength;
+
+    Status = USBH_AllocateWorkItem(HubExtension,
+                                   &HubWorkItem,
+                                   USBH_ChangeIndicationWorker,
+                                   BufferLength,
+                                   (PVOID *)&HubWorkItemBuffer,
+                                   DelayedWorkQueue);
+
+    if (!NT_SUCCESS(Status))
+    {
+        return STATUS_MORE_PROCESSING_REQUIRED;
+    }
+
+    RtlZeroMemory(HubWorkItemBuffer, BufferLength);
+
+    HubWorkItemBuffer->IsRequestErrors = FALSE;
+
+    if (IsErrors)
+    {
+        HubWorkItemBuffer->IsRequestErrors = TRUE;
+    }
+
+    if (InterlockedIncrement(&HubExtension->ResetRequestCount) == 1)
+    {
+        KeResetEvent(&HubExtension->ResetEvent);

```

Comment:
usbhub.c:2314: did you mean Port < NumPorts?

The bitmap uses 0-based port numbers, the function you call later presumably uses 1-based?
In that case, this would make more sense to me:
```
for (Port = 0; Port < NumPorts; Port++)
{
    if (IsBitSet(Bitmap, Port)
    {
        break;
    }
}
if (Port == NumPorts)
{
    Port = 0;
}
else
{
    Port++;
}
```

or

```
for (Port = 1; Port <= NumPorts; Port++)
{
    if (IsBitSet(Bitmap, Port - 1)
    {
        break;
    }
}
if (Port > NumPorts)
{
    Port = 0;
}
```